hivemind/client/averaging/__init__.py

```diff
@@ -324,35 +312,43 @@ async def _step(self, *, future: MPFuture, gather_binary: bytes, weight: float,
+            incoming_throughputs = [thr if mode != AveragingMode.CLIENT else 0.0
+                                    for thr, mode in zip(throughputs, modes)]
+            peer_fractions = await asyncio.get_event_loop().run_in_executor(
                 None, load_balance_peers, self.total_size, incoming_throughputs, min_vector_size)
-            async with self.get_tensors_async() as averaged_tensors:
-                return AllReduceRunner(group_id=group_info.group_id, tensors=averaged_tensors, endpoint=self.endpoint,
-                                       ordered_group_endpoints=group_info.endpoints, part_sizes=part_sizes,
-                                       weights=weights, gathered=user_gathered, return_deltas=True, modes=modes, **kwargs)
-        except Exception as e:
-            raise MatchmakingException(f"Unable to create allreduce runner ({e}), group_info: {weights, throughputs, modes, user_gathered}")
 
-    def update_tensors(self, allreduce_group: AllReduceRunner):
-        """
-        a private (extendable) method that applies changes from a finished allreduce to local tensors
-        """
-        assert allreduce_group.return_deltas and allreduce_group.future.done()
-        averaging_deltas = allreduce_group.future.result()
+            async with self.get_tensors_async() as local_tensors:
+                allreduce_protocol = AllReduceProtocol(
+                    group_id=group_info.group_id, tensors=local_tensors, endpoint=self.endpoint,
+                    ordered_group_endpoints=group_info.endpoints, peer_fractions=peer_fractions, weights=weights,
+                    gathered=user_gathered, modes=modes, **kwargs)
+                self._running_groups[group_info.group_id] = allreduce_protocol
+                self._pending_group_assembled.set()
 
-        with torch.no_grad(), self.get_tensors() as local_tensors:
-            assert len(local_tensors) == len(self._averaged_tensors)
-            for tensor, update in zip(local_tensors, averaging_deltas):
-                tensor.add_(update, alpha=self._averaging_alpha)
-        self.last_updated = get_dht_time()
+                # actually run all-reduce
+                averaging_outputs = [output async for output in allreduce_protocol]
+
+                if modes[group_info.endpoints.index(self.endpoint)] != AveragingMode.AUX:
+                    assert len(local_tensors) == len(self._averaged_tensors)
+                    for tensor, update in zip(local_tensors, averaging_outputs):
+                        tensor.add_(update, alpha=self._averaging_alpha)
+                    self.last_updated = get_dht_time()
+
+                return allreduce_protocol.gathered
+        except BaseException as e:
+            logger.exception(e)
+            raise MatchmakingException(f"Unable to run all-reduce ({e})")
+        finally:
+            _ = self._running_groups.pop(group_info.group_id, None)
```

Comment:
No need to assign a value

Reply:
it was required for pycharm, and it is still required by pycharm